### ISSUE TITLE
copyright agent: pre-filter for results, or split scanning of URLs, e-mail addresses, etc.

### ISSUE THREAD

**Issue by yaobinshi** [fossology/fossology#258]
---

Author Name: **armijn hemel** 
Original Redmine Issue: 6405, http://www.fossology.org/issues/6405
Original Date: 2014/01/15
Original Assignee: larry shi

---

When generating the database for the Binary Analysis Tool I use the FOSSology copyright agent to extract copyright information. This is quite costly: a large part of creation of the database is running the copyright agent. Even when running in parallel it takes quite a bit of time: for a recent Linux kernel it easily takes 10 minutes on a fast machine, with 8 copyright agents running in parallel.

The copyright agent currently can extract URLs, e-mail addresses and statements and uses regular expressions for this, which can be quite expensive. However, a lot of filtering can already be done beforehand. An e-mail address always contains the '@' character, aURL always contains '://' (at least according to the regular expression used).

I did some tests with Linux 3.9.6 to see how many files the copyright agent would find these character sequences for:

[armijn@lurleen linux-3.9.6]$ find . -name "*.c" | wc -l
18364

[armijn@lurleen linux-3.9.6]$ find . -name "*.c" | xargs grep '://' | cut -f 1 -d : | sort | uniq | wc -l
1788

[armijn@lurleen linux-3.9.6]$ find . -name "*.c" | xargs grep '@' | cut -f 1 -d : | sort | uniq | wc -l
13092

These results mean that in only 10% of the .c files in this version of the kernel the copyright agent will possibly find a URL. In about 2/3 of the .c files in the kernel the copyright agent will possibly find an e-mail address. Not having to run these regular expressions could mean a significant performance boost for the copyright agent when processing vast amounts of data (which I do on a regular basis).

There are two solutions for my problem here:

1. split the scanning into a few cases and let the user pick which cases to scan for using a commandline switch (I would do my own filtering then before even launching the copyright agent)

2. do some pre-filtering in the copyright agent to prevent the regular expression from being run

From a maintenance point of view the second solution is probably best.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **armijn hemel** 
Original Date: 2014/01/15 14:39:01 +0000

---

some more data:

[armijn@lurleen linux-3.9.6]$ find . -name "*.h" | wc -l
15306
[armijn@lurleen linux-3.9.6]$ find . -name "*.h" | xargs grep '@' | cut -f 1 -d : | sort | uniq | wc -l
6122
[armijn@lurleen linux-3.9.6]$ find . -name "*.h" | xargs grep '://' | cut -f 1 -d : | sort | uniq | wc -l
855

A small calculation:

>>> 100 - (6122+855)/(15306.0*2) * 100
77.20828433294133

So a bit more than 77% of the regular expressions that would be run for the .h files for this kernel would never give a URL or e-mail address match.

If needed I can get more data from my database to find out the hit rate of these regular expressions :-)

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **armijn hemel** 
Original Date: 2014/01/16 23:45:16 +0000

---

I made a crude patch and then I did a test on a dual core laptop scanning the copyrights of Linux kernel 2.4.10, plus a few other things for the BAT database (but 90% of the time is spent on copyright scanning).

My script using the standard version of the copyright agent:

real    12m17.636s
user    14m41.467s
sys     7m52.455s

The version of the copyright agent with the attached patch is slightly faster and results are the same (times verified with various runs to not get caught by the kernel or disk caching results):

real    11m57.814s
user    14m13.241s
sys     7m41.668s

So it seems that the URL and e-mail extraction are not the most expensive operations in the copyright agent, so if there is any time for further improvement, that would be the place to spend it.

I would appreciate it if you could review and (hopefully) apply the patch.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **armijn hemel** 
Original Date: 2014/01/17 00:04:28 +0000

---

Oh, if you apply it you might want to fix a few of the comments first. I just copied/pasted them for the original.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2014/01/17 06:45:18 +0000

---

copyright agent spends most time on extracting copyright result.

also I will compare the performance between copyright agent and commands on three parts,
1) extracting copyright information, copyright text position, r/w database
2) extracting url information, url text position, r/w database
3) extracting email information, email text position, r/w database

Armijn, how did you run copyright agent to extract copyright information for Linux kernel 2.4.10? I would like to reproduce your testing.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2014/01/17 06:51:52 +0000

---

> There are two solutions for my problem here:
> 
> 1. split the scanning into a few cases and let the user pick which cases to scan for using a commandline switch (I would do my own filtering then before even launching the copyright agent)
> 
> 2. do some pre-filtering in the copyright agent to prevent the regular expression from being run
> 
> From a maintenance point of view the second solution is probably best.
add a switch to scan for using a commandline is a good idea. I like it.
there is one thing, when using a commandline  extracting(not regular expression), store the extracting data into database or not? I mean you just want to copyright agent from with -C(not from the scheduler), will not store the extracting data into database, will save lots of work.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **armijn hemel** 
Original Date: 2014/01/17 10:24:48 +0000

---

The script I used can be found here:

http://www.binaryanalysis.org/trac/browser/trunk/src/maintenance/batchextractprogramstrings.py

Please note that there are some specifics you need to know before running it, contact me via normal e-mail if you run into issues.

On line 687 you can see that I run the copyright agent with the -C switch.

If my script is not practical I could also whip up a quick Python script that just runs the copyright agent in parallel (with a pool of workers) so you can focus on just the copyright agent. Let me know.

Regardless I think the patch might be worth applying, if you agree with it. Although in this example the performance gain is just 2.7% in my case that still means quite some hours.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **armijn hemel** 
Original Date: 2014/01/19 22:59:06 +0000

---

larry shi wrote:
 
> > From a maintenance point of view the second solution is probably best.
> add a switch to scan for using a commandline is a good idea. I like it.

If both methods could be used (so a switch so I can turn *off* statement scanning except URL and e-mail, as well as the patch that I proposed) that would make me very happy. At the moment I am mostly interested in URLs and e-mail addresses, not per se in copyright statements. Having a switch to turn it off and only get the information I want is great.

If you would do that, could you also do the following (if possible) read names.dic after the commandline options have been given to avoid unnecessarily reading the file if statement scanning is disabled? Remember that I run this a few dozen million times, so then it really has an impact :-)

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2014/01/21 09:17:53 +0000

---

armijn hemel wrote:
> larry shi wrote:
>  
> > > From a maintenance point of view the second solution is probably best.
> > add a switch to scan for using a commandline is a good idea. I like it.
> 
> If both methods could be used (so a switch so I can turn *off* statement scanning except URL and e-mail, as well as the patch that I proposed) that would make me very happy. At the moment I am mostly interested in URLs and e-mail addresses, not per se in copyright statements. Having a switch to turn it off and only get the information I want is great.
1) will try to add 2 methods to extract URL and e-mail.
2) will add options for statement extraction, URL extraction, e-mail extraction individually. Default as All extraction.
> 
> If you would do that, could you also do the following (if possible) read names.dic after the commandline options have been given to avoid unnecessarily reading the file if statement scanning is disabled? Remember that I run this a few dozen million times, so then it really has an impact :-)
sure.

all the requirements in this issue need a little more time to design and refactor our copyright agent, hope we can finish them in 2.5.0.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2014/03/07 04:30:48 +0000

---

have no time to finish this enhancement, shift it to next release 2.6.0

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2014/05/16 09:24:10 +0000

---

rough fix in git 213e76.

still load dict data even though do not want to get copyright statements, it is a little complicated. thinking.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2014/06/03 11:45:17 +0000

---


fixed in git 14999b. binary xxx 1st number as email(100), 2nd number as url(010), 3rd number as statement(001)

>read names.dic after the commandline options have been given to avoid unnecessarily reading the file if statement scanning is disabled?
I am sorry, avoid reading step will result in lots of code changes, please try if the new fix(git 14999b) can meet you needs, if still can not, I will pay more attention to it.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2014/07/28 09:00:42 +0000

---

how to verity this issue, please refer to http://www.fossology.org/projects/fossology/wiki/Copyright_Test_Cases#report-different-copyright-report-from-command-line

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **Bob Gobeille** 
Original Date: 2014/08/21 19:46:27 +0000

---

What is the result of your performance testing?  This whole issue is about improving performance.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **Bob Gobeille** 
Original Date: 2014/08/21 19:52:06 +0000

---

Using -T 1, -T2, -T7, ... is easy for the program but it is sure not friendly for users. 

What about something like:  -T C for copyrights, -T EUC for email+URL+copyrights?  The CEU order should not matter.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **larry shi** 
Original Date: 2014/08/22 08:01:58 +0000

---

#!/bin/bash
find /home/larry/fossology/src/nomos/agent_tests/testdata/NomosTestfiles/ $1 -type f -not \( -wholename "*svn*" \) -exec ./copyright -C  '{}' \;

~/fossology-2.5.0/src/copyright/agent$ time ./RunLicScan >log
real 2m10.017s 


with latest source code

time ./RunLicScan >log   
real 2m6.287s

in 2.6.0, the performance increase by 3%, not too much


time ./RunLicScan >log   -T 1
real 2m3.008s


time ./RunLicScan >log   -T 2
real 2m3.157s


 time ./RunLicScan >log   -T 4
real 2m3.016s


from the data above, you will find that, most time is spending on copyright initialization, I did not fix it in 2.6.0, that will require lots of code change.
possible we should put this issue into next release, or file another issue for the rest work of this issue.

what do you think?

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **Bob Gobeille** 
Original Date: 2014/08/22 19:02:58 +0000

---

Right 3% isn't much.  I also wonder how reproducible that is.   If it holds, 3% in a 24 hour scan will save 43 minutes.

* Comment #18 isn't clear but I'm guessing that you modified RunLicScan to use copyright (like you show in the first line).  Is that right?

* incidentally the real RunLicScan should be changed to use -d -n

* Moving this to 2.7 sounds like a good idea.  I don't think we need a new issue because it doesn't look like this issue is resolved.

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **armijn hemel** 
Original Date: 2014/08/22 19:15:18 +0000

---

I am really surprised that much of the time is actually spent for initialization. Earlier conversations seemed to suggest that it the copyright extraction bits would be the most resource hungry, but it turns out it's not. I think that this is a very interesting find.

3% might not seem much, but if you are processing a lot of files (and a machine is running for a full week) then it quickly adds up :-)

**Comment by yaobinshi** [comment]
---

Original Redmine Comment
Author Name: **Bob Gobeille** 
Original Date: 2014/08/22 21:08:47 +0000

---

Moving to 2.7.  Note that Larry did commit a change so the 3% is in there (assuming it is statistically significant).

**Comment by pguttmann** [comment]
Attachment - http://www.fossology.org/attachments/4497/copyright-2.4.0.patch

**Comment by maximilianhuber** [comment]
With the new C++ implementation, the copyright scan seems to be noticeably faster:
![tmp](https://cloud.githubusercontent.com/assets/1187050/11276460/744ae47c-8ee2-11e5-8b64-fc35a2c7e5ea.png)
(On notebook with current i7)
